File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}
	currentNode := &TreeNode{Val: preorder[0]}
	currentNodeIndex := 0
	for i := range inorder {
		if inorder[i] == currentNode.Val {
			currentNodeIndex = i
			break
		}
	}

	currentNode.Left = buildTree(preorder[1:currentNodeIndex+1], inorder[0:currentNodeIndex])
	currentNode.Right = buildTree(preorder[currentNodeIndex+1:len(preorder)], inorder[currentNodeIndex+1:len(inorder)])
	return currentNode
}

File: 105.construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder:
            return None
        root = TreeNode(preorder[0])
        rootInOrderIdx = inorder.index(root.val)
        leftInOrder = inorder[:rootInOrderIdx]
        rightInOrder = inorder[rootInOrderIdx + 1 :]
        leftPreOrder = preorder[1 : len(leftInOrder) + 1]
        rightPreOrder = preorder[len(leftInOrder) + 1 :]
        root.left = self.buildTree(leftPreOrder, leftInOrder)
        root.right = self.buildTree(rightPreOrder, rightInOrder)
        return root
